Add Lucene directory and index input implementations that expose shard snapshot

tlrx opened the pull request:
**Note:** this pull request targets the `feature/searchable-snapshots` branch

This pull requests adds a `SearchableSnapshotDirectory` implementation that exposes the snapshot of a shard as a Lucene `Directory`. 

This directory only supports read operations and does not allow any file modification. It allows:
- to list all the files stored in the shard snapshot using `Directory#listAll()`
- to return the byte length of a file in the snapshot using `Directory#fileLength(String)`
- to open an `IndexInput` for reading an existing file of the snapshot using `Directory#openInput(String, IOContext)`

In order to work, the directory requires the list of the shard snapshot files and a way to read a specific range of bytes blob. The list of shard snapshot files must be provided as a `BlobStoreIndexShardSnapshot` object when the directory is created. This object contains the list of the shard files stored in the snapshot and can be used to map each Lucene file with its corresponding blob(s) stored in the repository (which can be more than one as large Lucene files are split during snapshot). 

Blobs are directly read from the snapshot using a `BlobContainer`.

`SearchableSnapshotDirectory` provides `SearchableSnapshotIndexInput` to read a file from the snapshot. This index input implementation maintains an internal buffer (it extends `BufferedIndexInput`) and takes care of tracking current reading position in the file. Each time more bytes are requested to fill the internal buffer, `SearchableSnapshotIndexInput` maps the current position to the appropriate blob name and position in the blob to read bytes from. It also propagates the knowledge of the current position to any clone or slice.

This pull request also adds tests for the `SearchableSnapshotDirectory` which creates a random directory, index documents into it, snapshots the files and creates a `SearchableSnapshotDirectory` from this snapshot. It then runs some tests against the normal directory and the searchable snapshot directory and compares the results.

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Distributed)

tlrx commented:
@elasticmachine test this please

tlrx commented:
@DaveCTurner I adjusted this PR a bit so that the plugin provides a `DirectoryFactory` and the IndexInput now directly reads range of bytes using a `BlobContainer`. There are some `// NORELEASE` tags here and there but the whole logic works as expected and should unlock your work.

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public void close() throws IOException {
```
Nit:
```suggestion
    public void close() {
```

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public Set<String> getPendingDeletions() throws IOException {
```
Nit:

```suggestion
    public Set<String> getPendingDeletions() {
```

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public void sync(Collection<String> names) throws IOException {
```
Nit:
```suggestion
    public void sync(Collection<String> names) {
```

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public void syncMetaData() throws IOException {
```
Nit:
```suggestion
    public void syncMetaData() {
```

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public void deleteFile(String name) throws IOException {
```
Nit:

```suggestion
    public void deleteFile(String name) {
```

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
```
Nit:

```suggestion
    public IndexOutput createOutput(String name, IOContext context) {
```

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) throws IOException {
```
Nit:
```suggestion
    public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) {
```

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public void rename(String source, String dest) throws IOException {
```
Nit:

```suggestion
    public void rename(String source, String dest) {
```

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java:
```diff
@@ -0,0 +1,361 @@
+                .filter(file -> file.startsWith("extra") == false)
+                .forEach(file -> {
+                    try {
+                        assertThat(searchableDirectory.fileLength(file), equalTo(directory.fileLength(file)));
```
Could we add a message including the filename to this assertion?

tlrx replied:
Sure

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java:
```diff
@@ -0,0 +1,361 @@
+                    indexInput.seek(position);
+                    searchableIndexInput.seek(position);
+                }
+                assertThat("File pointers values should be the same before reading a byte",
```
Could we include `searchableIndexInput.toString()` and `indexInput.toString()` in the assertion message (and elsewhere in this test too).

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java:
```diff
@@ -0,0 +1,361 @@
+                        indexInput.seek(position);
+                        searchableIndexInput.seek(position);
+                    }
+                    assertThat("File pointers values should be the same before reading a byte",
```
Could we include `searchableIndexInput.toString()` and `indexInput.toString()` in the assertion message (and elsewhere in this test too).

tlrx replied:
Sure

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java:
```diff
@@ -0,0 +1,361 @@
+                ShardPath ephemeralShardPath = new ShardPath(false, tmpDir, tmpDir, ephemeralShardId);
+
+                final DirectoryFactory factory = SearchableSnapshots.newDirectoryFactory(() -> repositories);
+                try (Directory searchableDirectory = factory.newDirectory(ephemeralIndexSettings, ephemeralShardPath)) {
```
`searchable` seems like the wrong modifier here. Maybe `snapshotDirectory`? Also elsewhere in this test suite, and similarly for `searchableIndexInput`.

tlrx replied:
Sure

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:
```diff
@@ -5,11 +5,102 @@
+
+    @Override
+    public Map<String, DirectoryFactory> getDirectoryFactories() {
+        return Map.of("ephemeral", newDirectoryFactory(repositoriesService::get));
```
Makes sense, thanks

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:
```diff
@@ -5,11 +5,102 @@
+            }
+
+            BlobStoreRepository blobStoreRepository = (BlobStoreRepository) repository;
+            BlobContainer blobContainer = blobStoreRepository.shardContainer(new IndexId("_ephemeral",
```
Right

tlrx commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotIndexInputTests.java:
```diff
@@ -0,0 +1,149 @@
+            .thenAnswer(invocationOnMock -> {
+                String name = (String) invocationOnMock.getArguments()[0];
+                long position = (long) invocationOnMock.getArguments()[1];
+                int length = (int) invocationOnMock.getArguments()[2];
```
Sure

tlrx commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotIndexInputTests.java:
```diff
@@ -0,0 +1,149 @@
+
+                if (fileInfo.numberOfParts() == 1L) {
+                    if (name.equals(fileInfo.name()) == false || name.contains(".part")) {
+                        throw new IOException("Unexpected part name " + name);
```
I agree

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:
```diff
@@ -5,11 +5,102 @@
+            BlobContainer blobContainer = blobStoreRepository.shardContainer(new IndexId("_ephemeral",
+                shardPath.getShardId().getIndex().getUUID()), shardPath.getShardId().id());
+            BlobStoreIndexShardSnapshot snapshot = blobStoreRepository.loadShardSnapshot(blobContainer,
+                new SnapshotId("_ephemeral", EPHEMERAL_INDEX_SNAPSHOT_SETTING.get(indexSettings.getSettings())));
```
Sure

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:
```diff
@@ -5,11 +5,102 @@
+
+            BlobStoreRepository blobStoreRepository = (BlobStoreRepository) repository;
+            BlobContainer blobContainer = blobStoreRepository.shardContainer(new IndexId("_ephemeral",
+                shardPath.getShardId().getIndex().getUUID()), shardPath.getShardId().id());
```
Sure

tlrx commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java:
```diff
@@ -49,6 +49,22 @@
+     * @throws IOException         if the blob can not be read.
+     */
+    default InputStream readBlob(final String blobName, final long position, final int length) throws IOException {
+        throw new UnsupportedOperationException(); // NORELEASE
```
Yes, sorry if I wasn't clear. I did not include support for other blob store repositories like S3 because it adds more code to review and more plumbing. As I don't expect any difficulty I've just put aside this for now.

tlrx commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotIndexInputTests.java:
```diff
@@ -0,0 +1,149 @@
+                    break;
+                default:
+                    fail();
+            }
```
> This is a really nice piece of fuzz testing

I do agree, that's why I shamelessly borrow them from `ByteArrayIndexInputTests`.

> Maybe we should also occasionally seek elsewhere in the stream, read a few bytes, then seek back here again.

The `testReadByte` `testReadBytes` tests in SearchableSnapshotDirectoryTests do random seeks, but only in a forward manner, as Lucene "tries" to do. I think  the coverage is good enough?

DaveCTurner replied:
I think we are missing coverage of reads that aren't purely forwards and contiguous, i.e. an interleaving of seeks and reads (particularly given the three different read methods). For instance, reading from more than one place in the file, possibly jumping back-and-forth across blobs, seems like an interesting workload for a cache. To be clear I don't think there's a bug here, just that if I was going to add a bug here in future then that's where I think I'd put it :grin:

DaveCTurner replied:
a2de0d2815e71b4de4cab624d4ff1ca43dc0d466 looks good, thanks, although we'd better not be using the [XKCD random number generator](https://xkcd.com/221/)...


![](https://imgs.xkcd.com/comics/random_number.png)

tlrx replied:
That makes sense indeed. 

I added a change to the `randomReadAndSlice` method that seeks to a random position to read a single byte and then restore the previous position and pursue the execution (reading more than 1 byte would require more logic to handle corner cases though). It also overrides the byte read in the `output` value so that these extra reads are also verified as part of the test result.

tlrx commented:
@DaveCTurner Thanks for your feedback. I think I addressed all your comments. I don't expect the bwc tests to pass until the master is merged again in the feature branch (JDK13).

Can you please have another look?

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshots.java:
```diff
@@ -55,7 +61,7 @@ public SearchableSnapshots() {
     @Override
     public List<Setting<?>> getSettings() {
-        return List.of(EPHEMERAL_INDEX_REPOSITORY_SETTING, EPHEMERAL_INDEX_SNAPSHOT_SETTING);
+        return List.of(SNAPSHOT_REPOSITORY_SETTING, SNAPSHOT_SNAPSHOT_ID_SETTING, SNAPSHOT_INDEX_ID_SETTING);
```
Also `SNAPSHOT_SNAPSHOT_NAME_SETTING`?

tlrx replied:
Grrrr

Thanks

tlrx replied:
I pushed 8ad717425e704b5c235f20d6dd67ac45e2cd3e74

original-brownbear commented on server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:
```diff
@@ -163,6 +163,14 @@ public InputStream readBlob(String name) throws IOException {
 
+    @Override
+    public InputStream readBlob(String blobName, long position, int length) throws IOException {
+        final InputStream inputStream = readBlob(blobName);
```
I think you could just use this as the default implementation in `BlobContainer`  instead of throwing. I think all our streams support `skip`? :)

tlrx replied:
This is hacky and only less harmful for FS repository so I preferred to implement it in FS blob container and let the other implementation as not supported until the method is correctly implemented for each of them (using range of bytes download)

original-brownbear commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java:
```diff
@@ -49,6 +49,22 @@
+     * @throws NoSuchFileException if the blob does not exist
+     * @throws IOException         if the blob can not be read.
+     */
+    default InputStream readBlob(final String blobName, final long position, final int length) throws IOException {
```
Why would we want `length` on this API? Wouldn't it be better to just have `IndexInput` keep a reference to an open stream and only open a new stream if we seek backwards instead of opening a new stream of bounded length repeatedly?

tlrx replied:
It could be done like you suggest for FS repository (and maybe HDFS too) but for other repositories we need to give an indication of the number of bytes we want to download, because unlike the RestoreService we don't want to read all the blobs but only a chunk of it. Most SDK require to consume all the requested bytes (or will consume them under the hood for you) and we don't want to open a stream that reads a complete blob if we only use the first 28 bytes to read a header.

original-brownbear replied:
> Most SDK require to consume all the requested bytes (or will consume them under the hood for you) and we don't want to open a stream that reads a complete blob if we only use the first 28 bytes to read a header.

Fair point :) You have `abort` as a method on the S3 input stream though, not sure about GCS here.

original-brownbear commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java:
```diff
@@ -0,0 +1,147 @@
+    @Override
+    protected void readInternal(byte[] b, int offset, int length) throws IOException {
+        ensureOpen();
+        if (fileInfo.numberOfParts() == 1L) {
```
Why not reuse the logic from the restore codebase that has a sliced input stream already here instead of building the same thing again? (also see my comment on keeping a reference to a stream open until we're seeking backwards).

tlrx replied:
That's a good suggestion, we should be able to use SlicedInputStream combined with the length parameter. I'll take a look :)

tlrx commented:
@elasticmachine update branch

tlrx commented:
Thanks @DaveCTurner and @original-brownbear !

SherazT commented:
@tlrx Hey! I'm actually trying to parse snapshots `.dat` files and print the individual documents from them. I posted a question here (https://discuss.elastic.co/t/how-to-parse-snapshot-dat-file/218888) and I'm wondering if this feature does what I think it's doing? I have snap-{uuid}.dat files and have the further snap files relating to the shard, so will this feature help me print out the entries contained in the shard? Are these `files` that you refer you documents within each shard building up the snapshot? Would love your help on this.

tlrx commented:
@SherazT We replied on Discuss, it's better to keep the discussion there.
